### PR TITLE
Normative: when `regexp` arg is a string, minimize observable operations

### PR DESCRIPTION
**Pull request by ljharb**
Per #32.

cc @anba

### CODE REVIEW

File: spec.emu [hunk 1]
```diff
@@ -45,19 +41,29 @@ contributors: Jordan Harband
...
+			1. Let _flags_ be ? ToString(? Get(_R_, `"flags"`)).
+			1. Let _matcher_ be ? Construct(_C_, &laquo; _R_, _flags_ &raquo;).
+			1. Let _global_ be ? ToBoolean(? Get(_matcher_, `"global"`)).
+			1. Let _fullUnicode_ be ? ToBoolean(? Get(_matcher_, `"unicode"`).
+			1. Let _lastIndex_ be ? ToLength(? Get(_R_, `"lastIndex"`)).
+			1. Perform ? Set(_matcher_, `"lastIndex"`, _lastIndex_, *true*).
+		1. Else,
+			1. Let _flags_ be `"g"`.
+			1. Let _matcher_ be ? RegExpCreate(_regexp_, _flags_).
+			1. If ? IsRegExp(_matcher_) is not *true*, throw a *TypeError* exception.
```
Comment (anba) [inline spec.emu]: Why would you want to check `IsRegExp` after creating a regular expression object through `RegExpCreate`? I don't see any reason to care about programs which have deleted `RegExp.prototype[@@match]`.

File: spec.emu [hunk 2]
```diff
@@ -45,19 +41,29 @@ contributors: Jordan Harband
...
+			1. Let _global_ be *true*.
+			1. Let _fullUnicode_ be *false*.
+			1. Let _lastIndex_ be *0*.
+			1. Assert: ! Get(_matcher_, `"lastIndex"`) is *0*.
```
Comment (anba) [inline spec.emu]: You can't assert no abrupt completion occurs here, unless you remove `IsRegExp`. 😄

File: spec.emu [hunk 3]
```diff
@@ -45,19 +41,29 @@ contributors: Jordan Harband
...
 
 	<p>The abstract operation _MatchAllIterator_ performs the following steps:</p>
 	<emu-alg>
-		1. If ? IsRegExp(_R_) is not *true*, throw a *TypeError* exception.
+		1. Perform ! RequireObjectCoercible(_regexp_).
+		1. If ? IsRegExp(_regexp_) is `true`, then
+			1. Let _R_ be _regexp_.
+			1. Let _C_ be ? SpeciesConstructor(_R_, %RegExp%).
+			1. Let _flags_ be ? ToString(? Get(_R_, `"flags"`)).
+			1. Let _matcher_ be ? Construct(_C_, &laquo; _R_, _flags_ &raquo;).
```
Comment (anba) [inline spec.emu]: As mentioned in #32, I'd prefer `Let matcher be ? Construct(C, « R »).` here.

File: spec.emu [hunk 4]
```diff
@@ -16,14 +16,10 @@ contributors: Jordan Harband
 	<p>When the `matchAll` method is called, the following steps are taken:</p>
 	<emu-alg>
 		1. Let _O_ be ? RequireObjectCoercible(*this* value).
-		1. If ? IsRegExp(_regexp_) is *true*, then
-			1. Let _R_ be _regexp_.
-		1. Else,
-			1. Let _R_ be ? RegExpCreate(_regexp_, `"g"`).
-		1. Let _matcher_ be ? GetMethod(_R_, @@matchAll).
+		1. Let _matcher_ be ? GetMethod(_regexp_, @@matchAll).
```
Comment (anba) [inline spec.emu]: `String.prototype` methods, including `String.prototype.match`, generally allow `null`/`undefined` as parameters and implicitly coerce them to the expected type. Why should `String.prototype.matchAll` differ here?

File: spec.emu [hunk 5]
```diff
@@ -45,19 +41,29 @@ contributors: Jordan Harband
...
 <ins class="block">
 <emu-clause id="sec-matchalliterator" aoid="MatchAllIterator">
-	<h1>MatchAllIterator ( _R_, _O_ )</h1>
+	<h1>MatchAllIterator ( _regexp_, _O_ )</h1>
 
 	<p>The abstract operation _MatchAllIterator_ performs the following steps:</p>
 	<emu-alg>
-		1. If ? IsRegExp(_R_) is not *true*, throw a *TypeError* exception.
+		1. Perform ! RequireObjectCoercible(_regexp_).
+		1. If ? IsRegExp(_regexp_) is `true`, then
```
Comment (anba) [inline spec.emu]: `RegExp.prototype` methods, including `RegExp.prototype[@@match]`, generally throw a TypeError if the this-value is not an object. Why should `RegExp.prototype[@@matchAll]` differ here?

File: spec.emu [hunk 6]
```diff
@@ -45,19 +41,29 @@ contributors: Jordan Harband
...
+			1. Perform ? Set(_matcher_, `"lastIndex"`, _lastIndex_, *true*).
+		1. Else,
+			1. Let _flags_ be `"g"`.
+			1. Let _matcher_ be ? RegExpCreate(_regexp_, _flags_).
+			1. If ? IsRegExp(_matcher_) is not *true*, throw a *TypeError* exception.
+			1. Let _global_ be *true*.
+			1. Let _fullUnicode_ be *false*.
+			1. Let _lastIndex_ be *0*.
+			1. Assert: ! Get(_matcher_, `"lastIndex"`) is *0*.
 		1. Let _S_ be ? ToString(_O_).
```
Comment (anba) [inline spec.emu]: This should be moved into `String.prototype.matchAll` resp. `RegExp.prototype[@@matchAll]` for consistency with other `String.prototype` resp. `RegExp.prototype` methods.

File: spec.emu (hunk 1 above)
Comment (ljharb) [reply spec.emu]: I do; anything that's possible in an engine is something I care about.

File: spec.emu (hunk 5 above)
Comment (ljharb) [reply spec.emu]: It doesn't? `Perform ! RequireObjectCoercible(_regexp_).` throws a TypeError if `regexp` is not an object.

File: spec.emu (hunk 4 above)
Comment (ljharb) [reply spec.emu]: Good catch; this is missing a "neither undefined nor null, then" check.

File: spec.emu (hunk 2 above)
Comment (ljharb) [reply spec.emu]: can you elaborate on why not?

File: spec.emu (hunk 6 above)
Comment (ljharb) [reply spec.emu]: i'm not sure what you mean by "resp." here, can you elaborate?

File: spec.emu (hunk 3 above)
Comment (ljharb) [reply spec.emu]: as noted [here](https://github.com/tc39/proposal-string-matchall/issues/32#issuecomment-362440272), the fast path is explicitly and intentionally prohibited.

File: spec.emu (hunk 1 above)
Comment (littledan) [reply spec.emu]: Are there any other places in the specification with similar logic? There are many weird things that you can do with RegExp subclassing, but they don't tend to all have guards against them.

cc @allenwb

Comment (ljharb) [reply spec.emu]: Definitely! Here's all the RegExp-related places with fallback logic when the well-known symbol isn't callable:
 - https://tc39.github.io/ecma262/#sec-string.prototype.match
 - https://tc39.github.io/ecma262/#sec-string.prototype.split
 - https://tc39.github.io/ecma262/#sec-string.prototype.search
 - https://tc39.github.io/ecma262/#sec-isregexp
 - https://tc39.github.io/ecma262/#sec-string.prototype.replace

etc.

In other words, *every single one* of these methods has an identical fallback.

File: spec.emu (hunk 2 above)
Comment (anba) [reply spec.emu]:
```js
var RegExpPrototypeMatch = RegExp.prototype[Symbol.match];
Object.defineProperty(RegExp.prototype, Symbol.match, {
  // This getter function is called when the abstract operation IsRegExp is called.
  get() {
    this.lastIndex++;
    return RegExpPrototypeMatch;
  }
});
```

File: spec.emu (hunk 3 above)
Comment (anba) [reply spec.emu]: I don't understand the justification w.r.t `lastIndex`. Where does `lastIndex` come into play when 21.2.3.1, step 4 is taken?

Comment (anba) [reply spec.emu]: 1. Let C be ? SpeciesConstructor(R, %RegExp%).
1. Let flags be ? ToString(? Get(R, "flags")).
1. Let matcher be ? Construct(C, &laquo; R, flags &raquo;).
1. Let global be ? ToBoolean(? Get(matcher, "global")).
1. Let fullUnicode be ? ToBoolean(? Get(matcher, "unicode").

Needs the following guards to avoid creating a separate RegExp object:
- C is the built-in RegExp constructor
- R is a built-in RegExp object
- R doesn't redefine built-in RegExp.prototype functionality
- R.[[Prototype]] is the built-in RegExp.prototype object (Note: Ignoring subclassing for now.)
- RegExp.prototype.flags is the built-in getter
- Implied by RegExp.prototype.flags:
  - RegExp.prototype.global is the built-in getter
  - RegExp.prototype.ignoreCase is the built-in getter
  - RegExp.prototype.multiline is the built-in getter
  - RegExp.prototype.dotAll is the built-in getter
- RegExp.prototype.unicode is the built-in getter
- RegExp.prototype.sticky is the built-in getter
- RegExp.prototype.source is the built-in getter
- RegExp.prototype[@@ match] is the built-in function

If all those guards hold, we can directly access R.[[OriginalSource]] and R.[[OriginalFlags]] and save them in the RegExp String Iterator object. Then we also store R in the RegExp String Iterator object and for %RegExpStringIteratorPrototype%.next() we can reuse R for matching as long as the current R.[[OriginalSource]] and R.[[OriginalFlags]] match the original properties (that's why we need to store them in the RegExp String Iterator object!). When we reuse R in this case, the internal matching obviously needs to ensure that it doesn't modify the `lastIndex` property of R.

If we use the alternative:

1. Let C be ? SpeciesConstructor(R, %RegExp%).
1. Let matcher be ? Construct(C, « R »).
1. Let global be ? ToBoolean(? Get(matcher, "global")).
1. Let fullUnicode be ? ToBoolean(? Get(matcher, "unicode").

we need the following guards:
- C is the built-in RegExp constructor
- R is a built-in RegExp object
- R doesn't redefine built-in RegExp.prototype functionality
- R.[[Prototype]] is the built-in RegExp.prototype object (Note: Ignoring subclassing for now.)
- RegExp.prototype.global is the built-in getter
- RegExp.prototype.unicode is the built-in getter
- RegExp.prototype[@@ match] is the built-in function

So the alternative avoids the guards for RegExp.prototype.{global,ignoreCase,multiline,dotAll,source}. Does this make any difference for implementors? Well, it depends on the actual implementation. For example for SpiderMonkey we will only need to add a guard for RegExp.prototype.source to [1], the other getters are already checked. But I have no idea if it'll make any difference for other implementors, we'd need to ask them...

[1] https://searchfox.org/mozilla-central/rev/a539f8f7fe288d0b4d8d6a32510be20f52bb7a0f/js/src/builtin/RegExp.cpp#1614

File: spec.emu (hunk 2 above)
Comment (ljharb) [reply spec.emu]: In what way does that produce an abrupt completion? Or are you pointing out that at this point, a throwing getter could be defined on `this` for `lastIndex`?

File: spec.emu (hunk 3 above)
Comment (ljharb) [reply spec.emu]: In your second example, however, if *after* the iterator is created, any of those guards are violated, wouldn't you have to start down a much slower path than in the first example?

File: spec.emu (hunk 2 above)
Comment (anba) [reply spec.emu]: Oh, sorry I'm dumb. Pretend that I've written "You can't assert `lastIndex` is 0".

(Because if `RegExp.prototype[@@match]` has been redefined into a getter, it could have modified `lastIndex`.)

File: spec.emu (hunk 3 above)
Comment (anba) [reply spec.emu]: Updated the comment above to remove `RegExp.prototype.source` from the list of properties which need to guarded.

Comment (anba) [reply spec.emu]: For whatever reason I've always read 21.2.3.1, step 4 to require `flags` to be undefined... 😒 

So... 21.2.3.1, step 4 will basically always be taken in the fast path (where fast path means that `R` is a built-in RegExp object). 

> In your second example, however, if after the iterator is created, any of those guards are violated, wouldn't you have to start down a much slower path than in the first example?

No, I don't think so, because in both cases we create a fresh RegExp object (at least in the spec). 

The important part for me was if we can take 21.2.3.1, step 4, but given that I've misread if-condition in 21.2.3.1, my original concern was moot and we can probably leave it as is for now. As soon as we get some implementation experience, we may want to look at it again and see how good it can be optimized. 

I may also propose some other changes for spec consistency reasons, but I haven't yet decided which ones make more sense. For example `RegExp.prototype [ @@split ]` derives its `unicodeMatching` variable from the retrieved `flags` instead of getting the `unicode` property. So we may want to do the same `MatchAllIterator`. Or `RegExp.prototype [ @@match ]` and `RegExp.prototype [ @@replace ]` only retrieve `unicode` when `global` is true, whereas `MatchAllIterator` always gets the `unicode` property, which is another slight inconsistency.

File: spec.emu (hunk 2 above)
Comment (ljharb) [reply spec.emu]: ahh, right. that's fair.

Do you think this should throw if it's not zero? or do you instead think that the check should just be removed, and replaced with a runtime check (every time lastIndex is retrieved) to ensure that lastIndex is a valid index?

File: spec.emu (hunk 3 above)
Comment (ljharb) [reply spec.emu]: The retrieved flags are derived from checking each of the relevant properties: https://tc39.github.io/ecma262/#sec-get-regexp.prototype.flags

**Comment by ljharb** [comment]
It sounds like https://github.com/tc39/proposal-string-matchall/pull/33#discussion_r167389517 is the only remaining item on this PR; once that's resolved, it'd be great to get an approval so this can be merged :-)

Absolutely I'd love future issues to be filed if implementation experience indicates anything.

File: spec.emu (hunk 5 above)
Comment (anba) [reply spec.emu]: It doesn't cover `RegExp.prototype[@@matchAll].call(true, "")` which should throw a TypeError, because `true` is not an object.

**Comment by anba** [comment]
`"null".match(null)` returns `["null"]`, but `"null".matchAll(null)` throws a TypeError with the current patch.

File: spec.emu (hunk 1 above)
Comment (littledan) [reply spec.emu]: I don't see how those fallbacks are like this one. Those are used to determine whether to take the string path, not whether to throw a TypeError. The places I see IsRegExp used are more about throwing a TypeError when a RegExp is passed into a method intended for a string (and that's throwing if IsRegExp is true, not false),

**Comment by ljharb** [comment]
@anba with the current state of the PR, `'null'.matchAll(null)` should be identical to `'null'.matchAll('null')`, no?

**Comment by anba** [comment]
String.prototype.matchAll(_regexp_)
1. Let _O_ be ? RequireObjectCoercible(*this* value).
1. If _regexp_ is neither *undefined* nor *null*, then
    1. // not executed for _regexp_ = `null`
1. Return ? MatchAllIterator(_regexp_, _O_).


MatchAllIterator ( _regexp_, _O_ )
1. Perform ! RequireObjectCoercible(_regexp_). // <- Throws for _regexp_ = `null`

**Comment by ljharb** [comment]
Ah, thanks, you're right. In this case that's a bug, because the `!` is an assertion. Will fix.

**Comment by anba** [comment]
And `RegExp.prototype [ @@matchAll ] ( string )` should start with:

1. Let _rx_ be the **this** value.
1. If Type(_rx_) is not Object, throw a **TypeError** exception.
1. Return ? MatchAllIterator(_rx_, _string_).

**Comment by ljharb** [comment]
aha, thanks :-) that's why i'd added it originally. fixing that too.

**Comment by ljharb** [comment]
k; the PR is now updated; since the thread is quite long, can anyone (@littledan, @anba) post any remaining blockers for *this PR* - I would greatly appreciate further changes to be filed as separate issues, so I can consider those with a clean slate - as normal comments? (Linking to an existing review thread is totally great too)

**Comment by anba** [comment]
There are two nits in MatchAllIterator:
- `Let _R_ be _regexp_.` could be removed and instead of `R` the `regexp` parameter can be used directly.
- `Let _lastIndex_ be *0*.` can be removed, because `lastIndex` is unused in the else-branch.

Apart from that, I'd still like to see the `IsRegExp` call removed in the else-branch of MatchAllIterator.

And I'm not sure about the first `IsRegExp` call in MatchAllIterator and I wonder if it makes more sense to always take the first branch in MatchAllIterator if called from `RegExp.prototype[@@matchAll]` for consistency with `RegExp.prototype[@@split]`. But that should definitely be discussed in a different issue.

And finally, probably also for a different issue, the iterator calls `CreateIterResultObject(null, true)` instead of the normal `CreateIterResultObject(undefined, true)` when it is finished. I don't know if this is an intentional design decision or a bug?

**Comment by anba** [comment]
I've added some initial patches for SpiderMonkey at https://bugzilla.mozilla.org/show_bug.cgi?id=1435829.

**Comment by ljharb** [comment]
> There are two nits in MatchAllIterator:

Done.

> And I'm not sure about the first IsRegExp call in MatchAllIterator

#34

> And finally, probably also for a different issue, the iterator calls CreateIterResultObject(null, true) instead of the normal CreateIterResultObject(undefined, true) when it is finished. I don't know if this is an intentional design decision or a bug?

Unintentional; I've fixed that in master.

Thanks! If that takes care of all the issues for this PR, I'd appreciate a PR review approval :-D

File: spec.emu (hunk 2 above)
Comment (anba) [reply spec.emu]: > Do you think this should throw if it's not zero?

I think IsRegExp should be removed.

**Comment by anba** [comment]
> Thanks! If that takes care of all the issues for this PR, I'd appreciate a PR review approval :-D

There's still that issue with the IsRegExp call in the else-branch of MatchAllIterator. Do we also want to move the discussion for that one to a different PR?

**Comment by ljharb** [comment]
@anba I filed #34 to discuss that; are you comfortable merging this as-is, and discussing that change there?

**Comment by anba** [comment]
#34 is about the first IsRegExp call, but the other IsRegExp call is still problematic, as discussed in https://github.com/tc39/proposal-string-matchall/pull/33#discussion_r166351963 and https://github.com/tc39/proposal-string-matchall/pull/33#discussion_r166352273.

**Comment by ljharb** [comment]
I've changed the incorrect lastIndex assertion per https://github.com/tc39/proposal-string-matchall/pull/33#discussion_r166352273; could we move all "call IsRegExp less" discussion to #34? I'm happy to update its description.

**Comment by ljharb** [comment]
Thanks for all the feedback so far! I'm going to merge this; let's continue further review in separate issues/PRs.
